### ISSUE TITLE
3.0 Multiple contains with same name

### ISSUE THREAD

**Issue by Laykou** [cakephp/cakephp#5214]
Consider following table `Articles` with cached first and last comment:

```PHP
// Model/Table/ArticlesTable.php
$this->belongsTo( 'FirstComments', [
			'foreignKey' => 'first_comment_id',
			'className' => 'Comments'
			] );
$this->belongsTo( 'LastComments', [
			'foreignKey' => 'last_comment_id',
			'className' => 'Comments'
			] );

// Model/Table/CommentsTable.php
$this->belongsTo( 'Creators', [
			'foreignKey' => 'creator_id',
			'className' => 'Users'
			] );
```

When doing query

```PHP
$articles -> find() -> contain([ 'FirstComments' => ['Creators'], 'LastComments' => ['Creators'] ]);
```

it will not work, since there is twice **Creators** but in final query it is only once. The problem is that to make it fast it would be nice to have it just in **one query** instead of doing two queries. The main reason is also because **MSSQL** throws error when binding more than 2100 parameters (binding more than 2100 different ids).

I know there are already some solutions
- Add one more *belongsTo* binding named **Creators2** and use it like: ```$articles -> find() -> contain([ 'FirstComments' => ['Creators'], 'LastComments' => ['Creators2'] ]);```
- Use `'strategy' => 'subquery'`

But it would be nice if ORM would work just like I wrote it without any workarounds. Or are these *workarounds* recommended as a standart pattern?

**Comment by AD7six** [comment]
This is sort of related to https://github.com/cakephp/cakephp/pull/5911.

Another workaround would be to do this:

    // Model/Table/CommentsTable.php
    $this->belongsTo( $this->alias() . 'Creators', [
        'foreignKey' => 'creator_id',
        'className' => 'Users'
    ] );

I.e. - ensure that the association alias is unique. That would also create one instance of `Users` per association, instead of them all pointing at the same object.

I'm not sure if aliasing via contain is viable with the way things are currently implemented... then again by 3.1, jose might have found a way.

**Comment by TheLevti** [comment]
Hitting the same issue. I have a deeply nested query and because of repeating association names, the ideal query can not be constructed as I get `SQLSTATE[42S22]: Column not found: 1054 Unknown column 'AssocName.id' in 'field list'`. I figured, once the same association repeats, the first one is removed kind of when dumping the query.

Yes, one solution is to specify strategy to select, but the result is that for some reason it then in my case does 100 extra queries, each selecting a single entity only instead of collecting all ids and selecting all together.

Is the alias option within contain/matching already supported?
